src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
+                    if ((acc & ones) == ones)
+                    {
+                        // Is it a EOS. See: http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                        break;
+                    }
+                }
 
-            int codeMax = 0;
+                // Lookup index has to be 8 bits aligned to MSB
+                lookupIndex = (byte)(acc << (8 - bitsInAcc));
 
-            for (int i = 0; i < _decodingTable.Length && _decodingTable[i].codeLength <= validBits; i++)
-            {
-                (int codeLength, int[] codes) = _decodingTable[i];
+                ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
 
-                if (i > 0)
+                if (lookupValue < 0x80_00)
                 {
-                    codeMax <<= codeLength - _decodingTable[i - 1].codeLength;
-                }
+                    // Octet found.
+                    // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+                    // | 0 |     number_of_used_bits   |              octet            |
+                    // +---+---------------------------+-------------------------------+
+                    bitsInAcc -= lookupValue >> 8;
 
-                codeMax += codes.Length;
+                    if (bitsInAcc < 0)
+                    {
+                        // Last looked up code had more bits than was left in accumulator which indicated invalid or incomplete source
+                        throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
+                    }
 
-                int mask = int.MinValue >> (codeLength - 1);
-                long masked = (data & mask) >> (32 - codeLength);
+                    if (j == dst.Length)
+                    {
+                        Array.Resize(ref dstArray, dst.Length * 2);
+                        dst = dstArray;
+                    }
+                    dst[j++] = (byte)lookupValue;
 
-                if (masked < codeMax)
+                    // Set table index to root - start of new huffman code.
+                    lookupTableIndex = 0;
+                }
+                else
                 {
-                    decodedBits = codeLength;
-                    return codes[codes.Length - (codeMax - masked)];
+                    // Src was depleted in middle of lookup tree or EOS was decoded.
+                    throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
                 }
             }
 
-            decodedBits = 0;
-            return -1;
+            if (lookupTableIndex != 0)
+            {
+                // Finished in middle of traversing - no valid symbol could be decoded
+                // or too long EOS padding (7 bits plus). See: http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
```

Comment:
This exception is 4 times in this method, maybe move this to a throw-helper?

I.e.
```c#
private static void ThrowHuffmanDecodingException() => throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
```
and call this method instead of throwing directly.
This should move the exception-throwing into a cold part (machine code wise), and reduce the code size in `Decode`-method.

Reply:
I have benchmarked recommended changes and did not measured significant perf difference.
I'll keep it as is for code maintainability.,